.tree-sitter-lint/tree-sitter-lint-local/src/lib.rs

```diff
@@ -1 +1,126 @@
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+pub fn run_fixing_for_slice<'a>(
+    file_contents: impl Into<MutRopeOrSlice<'a>>,
+    tree: Option<&Tree>,
+    path: impl AsRef<Path>,
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_fixing_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+struct LocalLinterConcrete;
+
+impl LocalLinter for LocalLinterConcrete {
+    fn run_for_slice<'a>(
+        &self,
+        file_contents: impl Into<RopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_for_slice(file_contents, tree, path, args, language)
+    }
+
+    fn run_fixing_for_slice<'a>(
+        &self,
+        file_contents: impl Into<MutRopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_fixing_for_slice(file_contents, tree, path, args, language)
+    }
+}
+
+pub async fn run_lsp() {
+    lsp::run(LocalLinterConcrete).await;
+}
+
+fn args_to_config<T: FromFileRunContextInstanceProviderFactory>(args: Args) -> Config<T> {
+    args.load_config_file_and_into_config(all_plugins(), all_standalone_rules())
+}
+
+fn all_plugins<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Plugin<T>> {
+    vec![tree_sitter_lint_plugin_replace_foo_with::instantiate()]
+}
+
+fn all_standalone_rules<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Arc<dyn Rule<T>>> {
+    local_rules::get_rules()
+}
+
+struct FromFileRunContextInstanceProviderFactoryLocal;
+
+impl FromFileRunContextInstanceProviderFactory for FromFileRunContextInstanceProviderFactoryLocal {
+    type Provider<'a> = FromFileRunContextInstanceProviderLocal<'a>;
+
+    fn create<'a>(&self) -> Self::Provider<'a> {
+        FromFileRunContextInstanceProviderLocal {
+            tree_sitter_lint_plugin_replace_foo_with_provided_instances:
+                tree_sitter_lint_plugin_replace_foo_with::ProvidedTypes::<'a>::once_lock_storage(),
+        }
+    }
+}
+
+struct FromFileRunContextInstanceProviderLocal<'a> {
+    tree_sitter_lint_plugin_replace_foo_with_provided_instances:
+        <tree_sitter_lint_plugin_replace_foo_with::ProvidedTypes<'a> as FromFileRunContextProvidedTypes::<'a>>::OnceLockStorage,
+}
+
+impl<'a> FromFileRunContextInstanceProvider<'a> for FromFileRunContextInstanceProviderLocal<'a> {
+    type Parent = FromFileRunContextInstanceProviderFactoryLocal;
```

Comment:
The instance-provider trait points back to its factory "parent" (really this would probably be better named `Factory`?) since the `FileRunContext` struct (in the trait method signature below) is generic over the factory type so we need to be able to refer to it "from the child"